image-hd.c

```diff
@@ -411,6 +414,13 @@ static int hdimage_generate(struct image *image)
 	 * merely to silence a warn_unused_result warning.
 	 */
 	ret = truncate(imageoutfile(image), 0);
+	/*
+	 * Resize the file immediately to the final size. This is not
+	 * strictly necessary but this circumvents XFS preallocation
+	 * heuristics. Without this, the holes in the image may be smaller
+	 * than necessary.
+	 */
+	ret = truncate(imageoutfile(image), hd->file_size);
```

Comment:
Eh, what does that do in the case the output image file doesn't already exist?

Reply:
[`truncate()` returns -1, errno set to ENOENT](https://man.archlinux.org/man/truncate.2)

Reply:
True. And the test worked because in this case, the image was already there.

I expanded the test to ensue this is caught as well and fixed the code to explicitly create the file first.